File: internal/config/cycle.go
package config

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/BurntSushi/toml"
)

// ---------------------------------------------------------------------------
// Baukasten ("cycle") schema — stored per-cycle under the DATA dir so GUI edits
// survive: ~/.local/share/symvibe/cycles/<id>.toml
// ---------------------------------------------------------------------------

// Cycle is one editable Baukasten: an ordered set of phases, each with steps.
// The struct carries both toml (on-disk) and json (on-the-wire) tags using the
// SAME snake_case names so disk and GUI share one mental model.
type Cycle struct {
	SchemaVersion int     `toml:"schema_version" json:"schema_version"`
	ID            string  `toml:"id" json:"id"`
	Name          string  `toml:"name" json:"name"`
	Description   string  `toml:"description" json:"description"`
	Phases        []Phase `toml:"phases" json:"phases"`
}

// Phase is a category/column on the board.
type Phase struct {
	ID    string `toml:"id" json:"id"`
	Name  string `toml:"name" json:"name"`
	Order int    `toml:"order" json:"order"`
	Steps []Step `toml:"steps" json:"steps"`
}

// Step is one block. Skill binds it to an opencode skill/command; Category binds
// it to a model. ModelOverride, when non-nil, beats the category binding.
// Status is RUNTIME state, persisted so the cycle knows where it left off.
type Step struct {
	ID            string `toml:"id" json:"id"`
	Name          string `toml:"name" json:"name"`
	Order         int    `toml:"order" json:"order"`
	Skill         string `toml:"skill" json:"skill"`                 // opencode command/skill name
	Category      string `toml:"category" json:"category"`           // model binding key
	Agent         string `toml:"agent" json:"agent"`                 // optional opencode agent override
	PromptSuffix  string `toml:"prompt_suffix" json:"prompt_suffix"` // optional extra instruction
	Enabled       bool   `toml:"enabled" json:"enabled"`
	ModelOverride *Model `toml:"model_override" json:"model_override,omitempty"` // optional per-step model

	// AutoSkip is an optional data-driven skip rule. Before running the step the
	// engine evaluates the named Sensor; if its value satisfies When, the step is
	// marked skipped and the runner is never invoked (e.g. sensor="open-issues"
	// when="==0" -> skip when there are no open issues). This is how the cycle
	// "knows what to skip" without hard-coding it.
	AutoSkip *AutoSkip `toml:"auto_skip" json:"auto_skip,omitempty"`

	// DependsOn lists step IDs that must be terminal before this one may run.
	// Reserved for the engine; the MVP walks strictly in board order.
	DependsOn []string `toml:"depends_on" json:"depends_on,omitempty"`

	// ParallelSafe marks a step that may run concurrently with phase siblings.
	// Reserved; the MVP runs one step at a time on a shared working tree.
	ParallelSafe bool `toml:"parallel_safe" json:"parallel_safe,omitempty"`

	// Runtime status (persisted across runs so the cycle is resumable). An empty
	// value means pending — see StepStatus.Effective.
	Status StepStatus `toml:"status" json:"status"`
}

// AutoSkip is a cheap pre-step probe + predicate. The engine runs Sensor (a
// fast, side-effect-free probe such as git-dirty / open-issues / open-prs) and
// skips the step when the integer result satisfies When (e.g. "==0", ">0",
// ">=3", "!=0").
type AutoSkip struct {
	Sensor string `toml:"sensor" json:"sensor"`
	When   string `toml:"when" json:"when"`
}

// StepStatus drives the discreet GUI status icons. The wire/JSON and the
// frontend StepStatus union MUST use these exact string values.
type StepStatus string

const (
	StatusPending     StepStatus = "pending"
	StatusInProgress  StepStatus = "in_progress"
	StatusDone        StepStatus = "done"
	StatusSkipped     StepStatus = "skipped"
	StatusFailed      StepStatus = "failed"
	StatusBlocked     StepStatus = "blocked"
	StatusNeedsReview StepStatus = "needs_review"
)

// Effective normalizes the empty zero-value (a freshly seeded step with no
// status key in its TOML) to pending so it is runnable.
func (s StepStatus) Effective() StepStatus {
	if s == "" {
		return StatusPending
	}
	return s
}

// IsTerminal reports whether the autonomous walk may pass over this step
// because it is finished (done or skipped).
func (s StepStatus) IsTerminal() bool {
	switch s.Effective() {
	case StatusDone, StatusSkipped:
		return true
	}
	return false
}

// IsHalting reports a status that stops the autonomous walk and requires a
// human decision or a resume-reset: failed, blocked, needs_review, or a stale
// in_progress left behind by a crash.
func (s StepStatus) IsHalting() bool {
	switch s.Effective() {
	case StatusFailed, StatusBlocked, StatusNeedsReview, StatusInProgress:
		return true
	}
	return false
}

// LoadCycle reads a cycle by id from the data dir. If it does not exist and id
// is the seed name, it materializes the embedded seed first.
func LoadCycle(id string) (*Cycle, error) {
	path := filepath.Join(CyclesDir(), id+".toml")
	if _, err := os.Stat(path); os.IsNotExist(err) {
		if id == SeedCycleName {
			if err := materializeSeed(path); err != nil {
				return nil, err
			}
		} else {
			return nil, fmt.Errorf("cycle %q not found at %s", id, path)
		}
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read cycle %q: %w", id, err)
	}
	var c Cycle
	if err := toml.Unmarshal(data, &c); err != nil {
		return nil, fmt.Errorf("parse cycle %q: %w", id, err)
	}
	return &c, nil
}

// SaveCycle atomically writes a cycle back to the data dir (GUI edit path).
func SaveCycle(c *Cycle) error {
	if c.ID == "" {
		return fmt.Errorf("save cycle: empty id")
	}
	if err := os.MkdirAll(CyclesDir(), 0o755); err != nil {
		return err
	}
	path := filepath.Join(CyclesDir(), c.ID+".toml")
	tmp := path + ".tmp"
	f, err := os.Create(tmp)
	if err != nil {
		return err
	}
	if err := toml.NewEncoder(f).Encode(c); err != nil {
		f.Close()
		return err
	}
	if err := f.Close(); err != nil {
		return err
	}
	return os.Rename(tmp, path) // atomic replace
}

// materializeSeed copies the embedded seed Baukasten to the data dir on first run.
func materializeSeed(dest string) error {
	data, err := seedFS.ReadFile("seed/seed-cycle.toml")
	if err != nil {
		return fmt.Errorf("read embedded seed: %w", err)
	}
	if err := os.MkdirAll(filepath.Dir(dest), 0o755); err != nil {
		return err
	}
	return os.WriteFile(dest, data, 0o644)
}

// NextRunnable returns the first enabled, pending step in board order — this is
// how the autonomous cycle "knows where it is". Done/skipped steps are walked
// past; the walk HALTS (ok=false) at the first enabled step in a halting state
// (failed/blocked/needs_review/in_progress) so the cycle never silently runs
// past a problem and never re-selects a failed step forever. Call ResetStuck
// before the first walk of a run to clear any crash-stale in_progress.
// Returns ok=false when the cycle is complete or halted.
func (c *Cycle) NextRunnable() (phaseIdx, stepIdx int, ok bool) {
	for pi := range c.Phases {
		for si := range c.Phases[pi].Steps {
			s := c.Phases[pi].Steps[si]
			if !s.Enabled {
				continue
			}
			switch {
			case s.Status.IsTerminal():
				continue
			case s.Status.IsHalting():
				return 0, 0, false
			default: // pending
				return pi, si, true
			}
		}
	}
	return 0, 0, false
}

// FindStep returns pointers to the phase and step with the given step id, or
// (nil, nil) if not found. The pointers are into the cycle's slices so callers
// may mutate status in place before SaveCycle.
func (c *Cycle) FindStep(stepID string) (*Phase, *Step) {
	for pi := range c.Phases {
		for si := range c.Phases[pi].Steps {
			if c.Phases[pi].Steps[si].ID == stepID {
				return &c.Phases[pi], &c.Phases[pi].Steps[si]
			}
		}
	}
	return nil, nil
}

// ResetStuck flips any persisted in_progress step back to pending so a crash
// mid-run does not wedge the cycle on resume. Returns the number reset.
func (c *Cycle) ResetStuck() (n int) {
	for pi := range c.Phases {
		for si := range c.Phases[pi].Steps {
			if c.Phases[pi].Steps[si].Status == StatusInProgress {
				c.Phases[pi].Steps[si].Status = StatusPending
				n++
			}
		}
	}
	return n
}

File: client/Sources/SymvibeKit/Models/Cycle.swift
import Foundation

public enum StepStatus: String, Codable, Sendable, CaseIterable {
    case pending
    case inProgress = "in_progress"
    case done
    case skipped
    case failed
    case blocked
    case needsReview = "needs_review"

    public var effective: StepStatus { self }

    public var isTerminal: Bool {
        self == .done || self == .skipped
    }

    public var isHalting: Bool {
        switch self {
        case .failed, .blocked, .needsReview, .inProgress:
            true
        default:
            false
        }
    }

    /// Status glyph for display (matches README mapping: ○ ◐ ✓ – ✕ ⦸ !).
    public var glyph: String {
        switch self {
        case .pending: "○"
        case .inProgress: "◐"
        case .done: "✓"
        case .skipped: "–"
        case .failed: "✕"
        case .blocked: "⦸"
        case .needsReview: "!"
        }
    }
}

public struct AutoSkip: Codable, Sendable, Equatable {
    public var sensor: String
    public var when: String

    public init(sensor: String, when: String) {
        self.sensor = sensor
        self.when = when
    }
}

/// Declarative per-step risk gate (`requires_review` on the wire). The client
/// does not edit it, but it must round-trip so a full `PUT /api/cycle` does not
/// silently drop the rule.
public struct RequiresReview: Codable, Sendable, Equatable {
    public var when: String

    public init(when: String) {
        self.when = when
    }
}

public struct StepModelOverride: Codable, Sendable, Equatable {
    public var id: String
    public var temperature: Double?
    public var variant: String?
    public var fallbackModels: [String]?

    public init(id: String, temperature: Double? = nil, variant: String? = nil, fallbackModels: [String]? = nil) {
        self.id = id
        self.temperature = temperature
        self.variant = variant
        self.fallbackModels = fallbackModels
    }
}

public struct Step: Codable, Sendable, Identifiable, Equatable {
    public var id: String
    public var name: String
    public var order: Int
    public var skill: String
    public var category: String
    public var agent: String?
    public var promptSuffix: String?
    public var enabled: Bool
    public var modelOverride: StepModelOverride?
    public var backendOverride: String?
    public var autoSkip: AutoSkip?
    public var requiresReview: RequiresReview?
    public var dependsOn: [String]?
    public var parallelSafe: Bool?
    public var status: StepStatus

    public init(
        id: String,
        name: String,
        order: Int,
        skill: String,
        category: String,
        agent: String? = nil,
        promptSuffix: String? = nil,
        enabled: Bool = true,
        modelOverride: StepModelOverride? = nil,
        backendOverride: String? = nil,
        autoSkip: AutoSkip? = nil,
        requiresReview: RequiresReview? = nil,
        dependsOn: [String]? = nil,
        parallelSafe: Bool? = nil,
        status: StepStatus = .pending
    ) {
        self.id = id
        self.name = name
        self.order = order
        self.skill = skill
        self.category = category
        self.agent = agent
        self.promptSuffix = promptSuffix
        self.enabled = enabled
        self.modelOverride = modelOverride
        self.backendOverride = backendOverride
        self.autoSkip = autoSkip
        self.requiresReview = requiresReview
        self.dependsOn = dependsOn
        self.parallelSafe = parallelSafe
        self.status = status
    }

    /// Steps embedded in an exported template carry no runtime status and may
    /// omit most fields, so decode defensively — the same type serves the board
    /// and the import/library payloads.
    public init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decodeIfPresent(String.self, forKey: .id) ?? ""
        name = try c.decodeIfPresent(String.self, forKey: .name) ?? ""
        order = try c.decodeIfPresent(Int.self, forKey: .order) ?? 0
        skill = try c.decodeIfPresent(String.self, forKey: .skill) ?? ""
        category = try c.decodeIfPresent(String.self, forKey: .category) ?? ""
        agent = try c.decodeIfPresent(String.self, forKey: .agent)
        promptSuffix = try c.decodeIfPresent(String.self, forKey: .promptSuffix)
        enabled = try c.decodeIfPresent(Bool.self, forKey: .enabled) ?? true
        modelOverride = try c.decodeIfPresent(StepModelOverride.self, forKey: .modelOverride)
        backendOverride = try c.decodeIfPresent(String.self, forKey: .backendOverride)
        autoSkip = try c.decodeIfPresent(AutoSkip.self, forKey: .autoSkip)
        requiresReview = try c.decodeIfPresent(RequiresReview.self, forKey: .requiresReview)
        dependsOn = try c.decodeIfPresent([String].self, forKey: .dependsOn)
        parallelSafe = try c.decodeIfPresent(Bool.self, forKey: .parallelSafe)
        status = try c.decodeIfPresent(StepStatus.self, forKey: .status) ?? .pending
    }
}

public struct Phase: Codable, Sendable, Identifiable, Equatable {
    public var id: String
    public var name: String
    public var order: Int
    public var steps: [Step]

    public init(id: String, name: String, order: Int, steps: [Step]) {
        self.id = id
        self.name = name
        self.order = order
        self.steps = steps
    }

    public init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decodeIfPresent(String.self, forKey: .id) ?? ""
        name = try c.decodeIfPresent(String.self, forKey: .name) ?? ""
        order = try c.decodeIfPresent(Int.self, forKey: .order) ?? 0
        steps = try c.decodeIfPresent([Step].self, forKey: .steps) ?? []
    }
}

public struct Cycle: Codable, Sendable, Identifiable, Equatable {
    public var schemaVersion: Int
    public var id: String
    public var name: String
    public var description: String
    public var phases: [Phase]

    public init(schemaVersion: Int, id: String, name: String, description: String, phases: [Phase]) {
        self.schemaVersion = schemaVersion
        self.id = id
        self.name = name
        self.description = description
        self.phases = phases
    }

    public init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        schemaVersion = try c.decodeIfPresent(Int.self, forKey: .schemaVersion) ?? 1
        id = try c.decodeIfPresent(String.self, forKey: .id) ?? ""
        name = try c.decodeIfPresent(String.self, forKey: .name) ?? ""
        description = try c.decodeIfPresent(String.self, forKey: .description) ?? ""
        phases = try c.decodeIfPresent([Phase].self, forKey: .phases) ?? []
    }

    // MARK: - Lookup helpers

    public func step(id stepID: String) -> Step? {
        for phase in phases {
            if let match = phase.steps.first(where: { $0.id == stepID }) { return match }
        }
        return nil
    }

    public func phase(containing stepID: String) -> Phase? {
        phases.first { $0.steps.contains { $0.id == stepID } }
    }

    /// Replaces the step carrying the same id, in place. No-op if it is unknown.
    public mutating func replace(_ step: Step) {
        for phaseIdx in phases.indices {
            if let stepIdx = phases[phaseIdx].steps.firstIndex(where: { $0.id == step.id }) {
                phases[phaseIdx].steps[stepIdx] = step
                return
            }
        }
    }
}
